drivers/adc/adc_nrfx_saadc.c

```diff
@@ -490,29 +525,78 @@ static int check_buffer_size(const struct adc_sequence *sequence, uint8_t active
 	return 0;
 }
 
+static inline void single_ended_channel_cut_negative_sample(uint16_t channel_bit,
+							    uint8_t  single_ended_channels,
+							    int16_t  **sample)
+{
+	if ((channel_bit & single_ended_channels) && (*sample < 0)) {
+		**sample = 0;
+	}
+
+	(*sample)++;
+}
+
 static bool has_single_ended(const struct adc_sequence *sequence)
 {
 	return sequence->channels & m_data.single_ended_channels;
 }
 
-static void correct_single_ended(const struct adc_sequence *sequence, nrf_saadc_value_t *buffer)
+static void correct_single_ended(const struct adc_sequence *sequence,
+				 nrf_saadc_value_t         *buffer,
+				 uint16_t                  buffer_size)
 {
-	uint16_t channel_bit = BIT(0);
 	uint8_t selected_channels = sequence->channels;
 	uint8_t single_ended_channels = m_data.single_ended_channels;
 	int16_t *sample = (int16_t *)buffer;
 
-	while (channel_bit <= single_ended_channels) {
-		if (channel_bit & selected_channels) {
-			if ((channel_bit & single_ended_channels) && (*sample < 0)) {
-				*sample = 0;
+	for (uint16_t channel_bit = BIT(0); channel_bit <= single_ended_channels;
+		channel_bit <<= 1) {
+		if (!(channel_bit & selected_channels)) {
+			continue;
+		}
+
+		if (m_data.internal_timer_enabled) {
+			if (!(channel_bit & single_ended_channels)) {
+				continue;
 			}
 
-			sample++;
+			for (int i = 0; i < buffer_size; i++) {
+				if (sample[i] < 0) {
+					sample[i] = 0;
+				}
+			}
+		} else {
+			single_ended_channel_cut_negative_sample(channel_bit,
+								 single_ended_channels,
+								 &sample);
 		}
+	}
+}
 
-		channel_bit <<= 1;
+/* The internal timer runs at 16 MHz, so to convert the interval in microseconds
+ * to the internal timer CC value, we can use the formula:
+ * interval_cc = interval_us * 16 MHz
+ * where 16 MHz is the frequency of the internal timer.
+ *
+ * The maximum value for interval_cc is 2047, which corresponds to
+ * approximately 7816 Hz ~ 128us.
+ * The minimum value for interval_cc is depends on the SoC.
+ */
+static inline uint16_t interval_to_cc(uint32_t interval_us)
```

Comment:
```suggestion
static inline uint16_t interval_to_cc(uint16_t interval_us)
```
Since this function must  called with `interval_us > 0` and `interval_us <= ADC_INTERNAL_TIMER_INTERVAL_MAX_US` it would be good to assert it.
Then error checking is not necessary.